cdap-data-fabric/src/main/java/io/cdap/cdap/data2/metadata/lineage/LineageTable.java

```diff
@@ -66,6 +67,46 @@ public static LineageTable create(StructuredTableContext context) {
     return new LineageTable(context);
   }
 
+  /**
+   * Deletes all completed run records with a start time before {@code timeUpperBound}, throwing
+   * {@code IOException} if the delete operation fails.
+   * <p>
+   * Assumption is that lineage records are added only after run is completed, so we don't need to
+   * check completed run statuses.
+   * </p>
+   *
+   * <p>
+   * This function will not work with the in memory no-sql DB and will return
+   * {@code InvalidFieldException}. This is because of the way we need to query and will only be
+   * used in the managed instances controlled by a flag.
+   * </p>
+   *
+   * @param timeUpperBound is the end time before which all records should be deleted.
+   */
+  public void deleteCompletedLineageRecordsStartedBefore(Instant timeUpperBound)
+      throws IOException {
+    // While converting from Run we are using Millis hence we need to get epoch millis.
+    long maxTimeEpoch = timeUpperBound.toEpochMilli();
+    if (maxTimeEpoch == Long.MAX_VALUE) {
```

Comment:
nit: If I would do this check, I would disallow anything after `System.currentTimeMillis`. If somehow we get an error value, it can be big, but still not MAX_VALUE

Reply:
I added this check to keep it consistent with the run record delete function. I feel that check for anything after systime may not still be the right thing at least for this specific function in Silo we may still pass systime-1ms and delete everything.

I am ok to remove this check all together or add systime check if you feel strongly about it. I don't have a strong preference either ways.

Reply:
No specific preference. I just don't understand the purpose of the check. If it's about catching random input, then Long.MAX_VALUE is pretty unlikely.

Reply:
Removed the checks all together.